### PR TITLE
feat: IAM identity mapping management

### PR DESCRIPTION
**Pull request by hacker65536**
Use this argument  `iam_identity_mapping`  to configure aws-auth configmap.
The backend uses `eksctl create|delete iamidentitymapping`.

cloud i have your check it ?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -250,14 +251,93 @@ resource "eksctl_cluster" "vpcreuse1" {
...
+      groups   = iam_identity_mapping.value["groups"]
+    }
+  }
+}
+
+output aws_auth {
+  value = eksctl_cluster.myeks.aws_auth_configmap
+}
+```
+
```
Comment (mumoshu) [inline README.md:319]:
```suggestion
On each `terraform apply`, the provider compares the current `aws-auth` configmap against the desired configmap contents, and run `eksctl create iamidentitymapping` to create additional mappings and `eksctl delete iamidentitymapping` to delete redundant mappings.

You can confirm the result by running `eksctl get iamidentitymapping`:
```

File: README.md [hunk 2]
```diff
@@ -250,14 +251,93 @@ resource "eksctl_cluster" "vpcreuse1" {
...
 ```console
 > kubectl get no
 NAME                                      STATUS                     ROLES    AGE    VERSION
 ip-10-0-4-28.us-east-2.compute.internal   Ready,SchedulingDisabled   <none>   4d1h   v1.16.13-eks-ec92d4
 ip-10-0-5-72.us-east-2.compute.internal   Ready                      <none>   4d1h   v1.16.13-eks-ec92d4
 ```
 
+## Add aws-auth ConfigMap
+
+You can use `iam_identity_mapping` to grant additional AWS users or roles to operate eks cluster with  aws-auth ConfigMap.
```
Comment (mumoshu) [inline README.md:264]:
```suggestion
You can use `iam_identity_mapping` to grant additional AWS users or roles to operate the EKS cluster by letting the provider to update the `aws-auth` ConfigMap.

To get started, add one or more `iam_identity_mapping` block(s) like in the below example:
```

File: pkg/resource/cluster/cluster_create.go [hunk 3]
```diff
@@ -132,3 +137,99 @@ func doWriteKubeconfig(d ReadWrite, clusterName, region string) error {
 
 	return nil
 }
+
+func createIAMIdentityMapping(d *schema.ResourceData, cluster *Cluster) error {
+
```
Comment (mumoshu) [inline pkg/resource/cluster/cluster_create.go:142]:
```suggestion
```

File: pkg/resource/cluster/cluster_create.go [hunk 4]
```diff
@@ -132,3 +137,99 @@ func doWriteKubeconfig(d ReadWrite, clusterName, region string) error {
...
+	iams, err := runGetIAMIdentityMapping(cluster)
+
```
Comment (mumoshu) [inline pkg/resource/cluster/cluster_create.go:144]:
```suggestion
```

File: pkg/resource/cluster/cluster_create.go [hunk 5]
```diff
@@ -132,3 +137,99 @@ func doWriteKubeconfig(d ReadWrite, clusterName, region string) error {
...
+	if len(iams) == 0 {
+		log.Printf("no data from eksctl get iamidentitymapping")
+	} else {
+		if err := d.Set(KeyAWSAuthConfigMap, iams); err != nil {
+			return fmt.Errorf("set aws-auth-configmap from iamidentitymapping : %w", err)
+		}
+	}
+
+	if d.Get(KeyIAMIdentityMapping) != nil {
+
```
Comment (mumoshu) [inline pkg/resource/cluster/cluster_create.go:158]:
```suggestion
```

File: pkg/resource/cluster/cluster_create.go [hunk 6]
```diff
@@ -132,3 +137,99 @@ func doWriteKubeconfig(d ReadWrite, clusterName, region string) error {
...
+		values := d.Get(KeyIAMIdentityMapping).(*schema.Set)
+		if err := runCreateIAMIdentityMapping(values, cluster); err != nil {
+			return fmt.Errorf("creating create  imaidentitymapping command: %w", err)
+		}
+		if err := d.Set(KeyIAMIdentityMapping, values); err != nil {
```
Comment (mumoshu) [inline pkg/resource/cluster/cluster_create.go:163]:
```suggestion

		if err := d.Set(KeyIAMIdentityMapping, values); err != nil {
```

File: pkg/resource/cluster/cluster_create.go [hunk 7]
```diff
@@ -132,3 +137,99 @@ func doWriteKubeconfig(d ReadWrite, clusterName, region string) error {
...
+			g2 = append(g2, v.(string))
+		}
+		args = append(args, g2...)
+
+		cmd, err := newEksctlCommandWithAWSProfile(cluster, args...)
+
+		if err != nil {
+			return fmt.Errorf("creating create imaidentitymapping command: %w", err)
+		}
+		res, err := resource.Run(cmd)
```
Comment (mumoshu) [inline pkg/resource/cluster/cluster_create.go:199]:
```suggestion

		res, err := resource.Run(cmd)
```

File: pkg/resource/cluster/cluster_create.go [hunk 8]
```diff
@@ -132,3 +137,99 @@ func doWriteKubeconfig(d ReadWrite, clusterName, region string) error {
...
+		if err != nil {
+			return fmt.Errorf("running create imaidentitymapping command: %w", err)
+		}
+		log.Printf("eksctl creat iamidentitymapping response: %v", res)
```
Comment (mumoshu) [inline pkg/resource/cluster/cluster_create.go:203]:
```suggestion

		log.Printf("eksctl creat iamidentitymapping response: %v", res)
```

File: pkg/resource/cluster/cluster_create.go [hunk 9]
```diff
@@ -132,3 +137,99 @@ func doWriteKubeconfig(d ReadWrite, clusterName, region string) error {
...
+			"--arn",
+			ele["iamarn"].(string),
+		}
+
+		cmd, err := newEksctlCommandWithAWSProfile(cluster, args...)
+
+		if err != nil {
+			return fmt.Errorf("creating create imaidentitymapping command: %w", err)
+		}
+		res, err := resource.Run(cmd)
```
Comment (mumoshu) [inline pkg/resource/cluster/cluster_create.go:226]:
```suggestion

		res, err := resource.Run(cmd)
```

File: pkg/resource/cluster/cluster_create.go [hunk 10]
```diff
@@ -132,3 +137,99 @@ func doWriteKubeconfig(d ReadWrite, clusterName, region string) error {
...
+		if err != nil {
+			return fmt.Errorf("creating create  imaidentitymapping command: %w", err)
```
Comment (mumoshu) [inline pkg/resource/cluster/cluster_create.go:228]:
```suggestion
			return fmt.Errorf("creating create-iamidentitymapping command: %w", err)
```

File: pkg/resource/cluster/cluster_read.go [hunk 11]
```diff
@@ -64,6 +70,10 @@ func (m *Manager) readCluster(d ReadWrite) error {
 			}
 		}
 	}
+	//Do not use d.Set , only for debug.
```
Comment (mumoshu) [inline pkg/resource/cluster/cluster_read.go:73]:
```suggestion
```

Can we remove this comment as (I thought) calling `d.Set` on Read should be avoided as it resulted in nasty errors/warnings?

File: pkg/resource/cluster/cluster_read.go [hunk 12]
```diff
@@ -108,3 +118,68 @@ func (m *Manager) planCluster(d *DiffReadWrite) error {
 
 	return nil
 }
+
+func readIAMIdentityMapping(d ReadWrite, cluster *Cluster) error {
+
```
Comment (mumoshu) [inline pkg/resource/cluster/cluster_read.go:123]:
```suggestion
```

File: pkg/resource/cluster/cluster_read.go [hunk 13]
```diff
@@ -108,3 +118,68 @@ func (m *Manager) planCluster(d *DiffReadWrite) error {
...
+	// get iamidentitymapping
```
Comment (mumoshu) [inline pkg/resource/cluster/cluster_read.go:124]:
```suggestion
```

I think this is obvious from the good function name(runGetIAMIdentityMapping) 😃

File: pkg/resource/cluster/cluster_read.go [hunk 14]
```diff
@@ -108,3 +118,68 @@ func (m *Manager) planCluster(d *DiffReadWrite) error {
...
+	iams, err := runGetIAMIdentityMapping(cluster)
+
```
Comment (mumoshu) [inline pkg/resource/cluster/cluster_read.go:126]:
```suggestion
```

File: pkg/resource/cluster/cluster_read.go [hunk 15]
```diff
@@ -108,3 +118,68 @@ func (m *Manager) planCluster(d *DiffReadWrite) error {
...
+	if err != nil {
+		return fmt.Errorf("can not get iamidentitymapping from eks cluster: %w", err)
+	}
+
+	current := make([]map[string]interface{}, 0)
+
+	for _, v := range d.Get(KeyAWSAuthConfigMap).(*schema.Set).List() {
+		current = append(current, v.(map[string]interface{}))
+	}
+	// sort for diff
```
Comment (mumoshu) [inline pkg/resource/cluster/cluster_read.go:136]:
```suggestion

	// sort for diff
```

**Review by mumoshu — CHANGES_REQUESTED** [review]
@hacker65536 Thanks a lot for your efforts! Implementation and design LGTM. I've added comments mostly about code style and documentation.

Would you mind merging my suggestion before proceeding?

File: pkg/resource/cluster/cluster_update.go [hunk 16]
```diff
@@ -225,6 +225,22 @@ func (m *Manager) updateCluster(d *schema.ResourceData) error {
 
 	}
 
+	updateIAMIdentityMapping := func() func() error {
+		return func() error {
+			d.HasChange(KeyIAMIdentityMapping)
+			a, b := d.GetChange(KeyIAMIdentityMapping)
+
+			if err := runCreateIAMIdentityMapping(b.(*schema.Set).Difference(a.(*schema.Set)), cluster); err != nil {
+				return fmt.Errorf("DeleteIAMIdentityMapping Error: %v", err)
```
Comment (oraclef2k20) [inline pkg/resource/cluster/cluster_update.go:234]:
```suggestion
				return fmt.Errorf("CreateIAMIdentityMapping Error: %v", err)
```

File: README.md [hunk 17]
```diff
@@ -250,14 +251,97 @@ resource "eksctl_cluster" "vpcreuse1" {
...
+  - system:nodes
+  rolearn: arn:aws:iam::123456789012:role/eksctl-myeks-nodegroup-ng2-NodeInstanceRole-2IGYK2W51ZHJ
+  username: system:node:{{EC2PrivateDNSName}}
+- groups:
+  - system:masters
+  rolearn: arn:aws:iam::123456789012:role/admin-role
+  username: admin-role
+- groups:
+  - system:masters
+  rolearn: arn:aws:iam::123456789012:user/user-admin
```
Comment (oraclef2k20) [inline README.md:342]:
```suggestion
  userarn: arn:aws:iam::123456789012:user/user-admin
```

File: pkg/resource/cluster/cluster_update.go (hunk 16 above)
Comment (hacker65536) [inline pkg/resource/cluster/cluster_update.go:234]:
```suggestion
				return fmt.Errorf("CreateIAMIdentityMapping Error: %v", err)
```

File: README.md (hunk 17 above)
Comment (hacker65536) [inline README.md:342]:
```suggestion
  userarn: arn:aws:iam::123456789012:user/user-admin
```

**Review by mumoshu — APPROVED** [review]
LGTM. Thanks again for your contribution @hacker65536 ☺️

**Pull request merged**